README.md
<h1 align="center">
  📸 Code Scenery
</h1>

<p align="center">
  <img src="https://img.shields.io/badge/vscode%20version-v1.0%2B-blue.svg" alt="VS Code Version" />
  <img src="https://img.shields.io/badge/license-MIT-green.svg" alt="License" />
</p>

<p align="center">
  <strong>A Visual Studio Code extension to create code snapshots</strong>
</p>


</p>

## 📺 Demo

[![Video Title](https://i.ibb.co/HpkQMsY/image.png)](https://stream.new/v/rjvy4k2OVs1L8PTShNUcwjlL027EjHxQCywfgn9ncN01Y)



## 🚀 Features

- Easily create snapshots of your code files within Visual Studio Code.
- Save and manage multiple snapshots for different versions of your code.
- Quick access to compare snapshots and track changes over time.
- Simple and intuitive interface for snapshot management.

## 📖 Installation

1. Launch Visual Studio Code.
2. Go to the Extensions view by clicking on the square icon on the left sidebar or pressing `Ctrl+Shift+X`.
3. Search for "Code Scenery" in the Extensions view search box.
4. Click on the "Install" button for the "Code Scenery" extension by "Mohamed Sawah".
5. Restart Visual Studio Code (if required).

## 🎯 Usage

1. Open the file you want to create a snapshot of.
2. Use the keyboard shortcut `Ctrl+Shift+P` to open the Command Palette.
3. Type "Code Scenery: Create Snapshot" and select it from the list.
4. Provide a name for the snapshot and press Enter.
5. The snapshot will be saved and added to the snapshot list.
6. To view or manage your snapshots, click on the "Code Snapshots" icon in the sidebar.

## 🤝 Contributing

Contributions are welcome! Here's how you can contribute:

1. Fork the repository and clone it to your local machine.
2. Create a new branch for your feature or bug fix.
3. Make your changes and test them thoroughly.
4. Commit your changes and push them to your forked repository.
5. Open a pull request, describing your changes in detail.

## 📄 License

This project is licensed under the [MIT License](LICENSE).

---


For any questions or issues, please [open an issue](https://github.com/MohamedSawahZC/code_scenery/issues) on GitHub.


---

### Live

<a href="https://marketplace.visualstudio.com/items?itemName=MohamedSawah.codescenery">VS Code Store</a>

CHANGELOG.md
# Change Log

All notable changes to the "codescenery" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';

const VIEW_TYPE = 'codescenery';
const WEB_VIEW_TITLE = 'Code Scenery';

let panel: vscode.WebviewPanel | undefined;

const init = (context: vscode.ExtensionContext) => {
    const activeTextEditor = vscode.window.activeTextEditor;
    if (activeTextEditor) {
        // Check if panel is already created, if not, create a new panel
        if (!panel) {
            panel = createPanel(context);
            // Dispose panel and clean up when closed
            panel.onDidDispose(() => {
                panel = undefined;
                vscode.window.showInformationMessage("See you later ❤️");
            });
        }
        // If there is text selected, update the panel
        if (hasTextSelected(activeTextEditor.selection)) {
            update(panel);
        }
    } else {
        //@desc Handle no text selection
        vscode.window.showErrorMessage("There is no text selected");
    }
};

const createPanel = (context: vscode.ExtensionContext): vscode.WebviewPanel => {
    const htmlTemplatePath = path.resolve(context.extensionPath, 'webview/index.html');
    const iconPath = path.resolve(context.extensionPath, 'webview/assets/images/icon-label.png');
    // Create a new webview panel
    const panel = vscode.window.createWebviewPanel(
        VIEW_TYPE,
        WEB_VIEW_TITLE,
        vscode.ViewColumn.Two,
        {
            enableScripts: true,
            localResourceRoots: [vscode.Uri.file(context.extensionPath)]
        }
    );

    panel.iconPath = vscode.Uri.file(iconPath);

    // Load HTML template into the webview panel
    panel.webview.html = getTemplate(htmlTemplatePath, panel);

    // Handle messages received from the webview
    panel.webview.onDidReceiveMessage((message) => {
        if (message.type === 'updateCode') {
            update(panel);
        }
    });

    return panel;
};

const getTemplate = (htmlTemplatePath: string, panel: vscode.WebviewPanel): string => {
    const htmlContent = fs.readFileSync(htmlTemplatePath, 'utf-8');
    // Replace placeholders in the HTML template with actual values
    return htmlContent
        .replace(/%CSP_SOURCE%/gu, panel.webview.cspSource)
        .replace(/(src|href)="([^"]*)"/gu, (_, match, src) => {
            let assetsPath = panel.webview.asWebviewUri(
                vscode.Uri.file(path.resolve(htmlTemplatePath, '..', src))
            );
            return `${match}="${assetsPath}"`;
        });
};

const update = (panel: vscode.WebviewPanel): void => {
    vscode.commands.executeCommand('editor.action.clipboardCopyAction');

    // Send a message to the webview to trigger the code update
    panel.webview.postMessage({
        type: 'updateCode'
    });
};

const hasTextSelected = (selection: vscode.Selection | undefined): boolean =>
    !!selection && !selection.isEmpty;

export const activate = (context: vscode.ExtensionContext) => {
    return context.subscriptions.push(
        // Register the extension command to capture code scenery
        vscode.commands.registerCommand('codescenery.capture', () => init(context))
    );
};


webview/assets/js/main.js
import { pasteCode, showLineNumbers, hideLineNumbers } from './code.js';
import { takeSnapshot } from './snapshot.js';

(() => {
    const snapshotContainerNode = document.querySelector('.snapshot-container');
    const snapshotContainerBackgroundNode = document.querySelector(
        '.snapshot-container__background'
    );
    const terminalNode = document.querySelector('.terminal');
    const sizeNode = document.querySelector('.header__size');
    const shootNode = document.querySelector('.shoot');
    const showLineNumbersNode = document.querySelector('#show-line-numbers');

    window.addEventListener('message', ({ data: { type } }) => {
        switch (type) {
            case 'updateCode':
                document.execCommand('paste');
                if (!showLineNumbersNode.checked ){
                    hideLineNumbers();
                }
                break;
        }
    });

    document.addEventListener('paste', event => {
        pasteCode(event.clipboardData);
    });

    shootNode.addEventListener('click', event => {
        takeSnapshot();
    });

    showLineNumbersNode.addEventListener('change', event => {
        const checkbox = event.target;

        if (checkbox.checked) {
            showLineNumbers();
        } else {
            hideLineNumbers();
        }
    });

    window.addEventListener('colorPickerChange', function(data) {
        const color = data.detail.el.value;
        snapshotContainerBackgroundNode.style.backgroundColor = color;
    });

    colorPicker.initAll();

    if (ResizeObserver) {
        const resizeObserver = new ResizeObserver(() => {
            let width = Math.round(snapshotContainerNode.offsetWidth) * 2;
            let height = Math.round(snapshotContainerNode.offsetHeight) * 2;

            sizeNode.textContent = width + 'x' + height;
        });

        resizeObserver.observe(snapshotContainerNode);
        resizeObserver.observe(terminalNode);
    }
})();


webview/assets/js/snapshot.js
const snapshotContainerNode = document.querySelector('.snapshot-container');
const snapshotContainerBackgroundNode = document.querySelector('.snapshot-container__background');
const terminalNode = document.querySelector('.terminal');
const exportFormatNode = document.getElementById('export-format');

export const takeSnapshot = (filename) => {
  snapshotContainerNode.style.resize = 'none';
  terminalNode.style.resize = 'none';

  const resetStyles = () => {
    snapshotContainerNode.style.resize = '';
    terminalNode.style.resize = '';
  };

  const options = {
    width: snapshotContainerBackgroundNode.offsetWidth * 2,
    height: snapshotContainerBackgroundNode.offsetHeight * 2,
    style: {
      transform: 'scale(2)',
      transformOrigin: 'center',
      background: '#e0eafc',
      background: 'linear-gradient(to left, #e0eafc, #cfdef3)'
    }
  };

  const selectedValue = exportFormatNode.value;

  if (selectedValue === 'svg') {
    domtoimage.toSvg(snapshotContainerBackgroundNode, options)
      .then(function (dataUrl) {
        resetStyles();
        const link = document.createElement('a');
        link.download = `codescenery-${filename}.svg`;
        link.href = dataUrl;
        link.click();
      });
  } else if(selectedValue === 'png') {
    domtoimage.toBlob(snapshotContainerBackgroundNode, options)
      .then(function (blob) {
        resetStyles();
        window.saveAs(blob, `codescenery-${filename}.${selectedValue}`);
      });
  }else{
    domtoimage.toBlob(snapshotContainerBackgroundNode, options)
      .then(function (blob) {
        resetStyles();
        window.saveAs(blob, `codescenery-${filename}.jpg`);
      });
  }
};


webview/assets/js/colorPicker.js
document.body.insertAdjacentHTML('afterbegin', '<div id="clpi-m"><div id="clpi-p"><div id="clpi-w"><div id="clpi-a"><div id="clpi-o"></div></div><div class="clpi-r"><div class="clpi-e clpi-e-hsl"> <input type="range" id="clpi-i-hsl" min="0" max="1" value="1" step="0.01"><div class="clpi-c" id="clpi-c-hsl"></div></div><div class="clpi-e" id="clpi-e-alpha"> <input type="range" id="clpi-i-alpha" min="0" max="1" value="1" step="0.01"><div class="clpi-c" id="clpi-c-alpha"></div></div></div></div><div class="clpi-b"><div class="clpi-b-hex"><input type="text" id="clpi-c"></div><div class="clpi-u"></div></div><div class="clpi-x"><div class="clpi-z"><div class="clpi-v"><div class="clpi-o" style="background: rgb(0, 0, 0);"></div><div class="clpi-o" style="background: rgb(255, 255, 255);"></div></div><div id="clpi-l"></div></div><div class="clpi-l-angle"> <input type="number" id="clpi-l-angle" placeholder="0"></div></div></div></div>');

var colorPicker = {
    input: null,
    thumb: null,
    value: '',
    h: 0,
    s: 0,
    l: 0,
    alpha: 1,

    gradientActive: false,
    gradientControl: null,
    gradientAngle: 0,
    gradientColor: document.getElementById('clpi-l'),
    gradientAngleInut: document.getElementById('clpi-l-angle'),

    mask: document.getElementById('clpi-m'),
    picker: document.getElementById('clpi-p'),

    currentColor: document.getElementById('clpi-c'),

    area: document.getElementById('clpi-a'),
    areaPosition: 0,

    pointer: document.getElementById('clpi-o'),
    pointerMove: false,

    hslInput: document.getElementById('clpi-i-hsl'),
    hslRangeControl: document.getElementById('clpi-c-hsl'),
    alphaRange: document.getElementById('clpi-e-alpha'),
    alphaInput: document.getElementById('clpi-i-alpha'),
    alphaRangeControl: document.getElementById('clpi-c-alpha'),

    cssClassControlActive: 'clpi-o-active',

    init: function (node) {
        let parent = node.closest('.clpi-wrapper'),
            input = parent.querySelector('input'),
            thumb = parent.querySelector('div');

        this.input = input;
        this.thumb = thumb;

        if (input.matches('.clpi-gradient')) {
            this.picker.classList.add('clpi-l-support');
        }

        this.mask.classList.add('clpi-m-show');

        let pos = input.getBoundingClientRect(),
            height = this.picker.getBoundingClientRect().height;

        let setToBottom = pos.top + pos.height + height > window.innerHeight;

        if (setToBottom) {
            setToBottom = pos.top - height - 5;
        } else {
            setToBottom = pos.top + pos.height + 5;
        }

        this.picker.style.top = setToBottom + 'px';
        this.picker.style.left = pos.left + 'px';

        let val = input.value;

        if (val.length < 3) return;

        if (val.includes('gradient')) {
            this.initGradient(val);
        } else if (val.includes('rgb')) {
            this.initRgb(val);
        } else {
            this.initHex(val);
        }
    },

    initGradient: function (val) {
        this.gradientActive = true;
        this.picker.classList.add('clpi-l-active');

        val = val.replace('linear-gradient(', '').slice(0, -1);

        let angle = val.split(',');
        angle = angle[0].replace(/[^\d,]/g, '');

        if (angle) {
            this.gradientAngleInut.value = parseInt(angle);
            this.gradientAngle = parseInt(angle);
        }

        let controls = val.match(/\((.*?)\)/g);

        // in case custom gradient input
        if (!controls) {return;}

        controls.forEach(function (v, k) {

            let rgb = v.replace(/[^\d,]/g, '').split(',').map(function (v) {
                return parseFloat(v);
            });

            if (rgb.length > 3) {
                rgb[3] = rgb[3] / 100;
            }

            document.querySelectorAll('.clpi-o')[k].style.background = rgb.length > 3 ? 'rgba(' + rgb.join(',') + ')' : 'rgb(' + rgb.join(',') + ')';

            controls[k] = rgb;
        });

        this.setGradientControl(document.querySelector('.clpi-o'));
    },

    initRgb: function (val) {
        let rgb = val.replace(/[^\d,]/g, '').split(',').map(function (v) {
            return parseFloat(v);
        });

        if (rgb.length > 3) {
            rgb[3] = rgb[3] / 100;
        }

        this.update(rgb);
    },

    initHex: function (val) {
        this.setCurrent(val);
    },

    change: function () {
        let n = this.hslRgb(this.h, this.s, this.l),
            hex = this.rgbHex(n);

        this.currentColor.value = hex;
        this.area.style.background = 'hsl(' + parseInt(this.h * 360) + ', 100%, 50%)';
        this.alphaRange.style.background = 'linear-gradient(to right, rgba(' + n.join(',') + ', 0) 0%, rgb(' + n.join(',') + ') 100%)';

        if (this.alpha === 1) {
            this.value = hex;
        } else {
            this.value = 'rgba(' + n.join(',') + ',' + this.alpha + ' )';
        }

        if (this.gradientActive) {
            this.gradientControl.style.background = this.value;

            let grColors = [];

            for (var item of document.querySelectorAll('.clpi-o')) {
                let bg = item.style.background;
                bg = bg ? bg : '#000';

                // firefox fix
                bg = bg.replace(' none repeat scroll 0% 0%', '');

                grColors.push(bg);
            }

            this.gradientColor.style.background = 'linear-gradient(to right, ' + grColors[0] + ' 0%, ' + grColors[1] + ' 100%)';

            let angle = this.gradientAngle ? this.gradientAngle + 'deg' : 'to bottom';

            this.value = 'linear-gradient(' + angle + ', ' + grColors[0] + ' 0%, ' + grColors[1] + ' 100%)';

        }

        this.thumb.style.background = this.value;

        this.input.value = this.value;
      
        let that = this;
      
        window.dispatchEvent(new CustomEvent('colorPickerTick', {
            detail: {
                el: that.input,
            }
        }));
    },

    triggerChange: function() {
        let that = this;

        window.dispatchEvent(new CustomEvent('colorPickerChange', {
            detail: {
                el: that.input,
            }
        }));
    },

    update: function (rgb) {
        let hsl = this.rgbHsl(rgb[0], rgb[1], rgb[2]);

        this.hslInput.value = hsl[0];

        this.setHslRange(hsl[0]);

        if (rgb.length > 3) {
            this.alphaInput.value = rgb[3];
            this.alpha = rgb[3];
            this.setAlphaRange(rgb[3]);
        } else {
            this.alphaInput.value = 1;
            this.alpha = 1;
            this.setAlphaRange(1);
        }

        this.h = hsl[0];
        this.s = hsl[1];
        this.l = hsl[2];

        hsl[2] = hsl[2] * 200 - 200;
        hsl[2] = hsl[2] < 0 ? hsl[2] * -1 : hsl[2];

        this.pointer.style.top = hsl[2] + 'px';
        this.pointer.style.left = hsl[1] * 200 + 'px';

        this.change();
    },

    close: function () {
        for (var item of document.querySelectorAll('.clpi-o')) {
            item.classList.remove(this.cssClassControlActive);
        }

        this.mask.classList.remove('clpi-m-show');
        this.picker.classList.remove('clpi-l-active');
        this.picker.classList.remove('clpi-l-support');

        this.reset();
    },

    reset: function () {
        this.h = 0;
        this.s = 0;
        this.l = 0;
        this.alpha = 1;
        this.value = '';

        this.gradientActive = false;
        this.gradientAngle = 0;

        this.pointer.removeAttribute('style');
        this.alphaRangeControl.removeAttribute('style');
        this.hslRangeControl.removeAttribute('style');
        this.currentColor.value = '';
        this.gradientAngleInut.value = '';
        this.hslInput.value = 1;
        this.alphaInput.value = 1;
    },

    setHsl: function (v) {
        this.h = v;
        this.setHslRange(v);
        this.change();
    },

    setHslRange: function (v) {
        this.hslRangeControl.style.left = v * 100 + '%';
    },

    setAlpha: function (v) {
        this.alpha = v;
        this.setAlphaRange(v);
        this.change();
    },

    setAlphaRange: function (v) {
        this.alphaRangeControl.style.left = v * 100 + '%';
    },

    setCurrent: function (hex) {
        if (hex.length < 6) {return;}

        let rgb = this.hexRgb(hex);

        if (!rgb) {return;}

        this.update(rgb);

        this.triggerChange();
    },

    setGradientControl: function (e) {
        this.gradientControl = e;

        for (var item of document.querySelectorAll('.clpi-o')) {
            item.classList.remove(this.cssClassControlActive);
        }

        this.gradientControl.classList.add(this.cssClassControlActive);


        let bg = this.gradientControl.style.background;

        if (bg) {
            // firefox fix
            bg = bg.replace(' none repeat scroll 0% 0%', '');
            let rgb = bg.replace(/[^\d,]/g, '').split(',').map(function (v) {
                return parseFloat(v);
            });

            if (typeof rgb[3] !== 'undefined') {
                rgb[3] = rgb[3] / 100;
            }

            this.update(rgb);
        }
    },

    setGradientAngle: function (v) {
        this.gradientAngle = v;
        this.change();
        this.triggerChange();
    },

    gradientShowToggle: function () {
        for (var item of document.querySelectorAll('.clpi-o')) {
            item.classList.remove(this.cssClassControlActive);
        }

        this.picker.classList.toggle('clpi-l-active');
        this.gradientActive = !this.gradientActive;

        if (this.gradientActive) {
            this.gradientControl = document.querySelector('.clpi-o');
            this.gradientControl.classList.add(this.cssClassControlActive);
        }

        this.change();
        this.triggerChange();
    },

    startDrag: function (e) {
        this.pointerMove = true;
        this.areaPosition = this.area.getBoundingClientRect();
        this.drag(e);
    },

    drag: function (e) {
        if (this.pointerMove) {
            let y = e.clientY - this.areaPosition.top,
                x = e.clientX - this.areaPosition.left;

            y = y < 0 ? 0 : y;
            y = y > 200 ? 200 : y;

            x = x < 0 ? 0 : x;
            x = x > 200 ? 200 : x;

            this.pointer.style.top = y + 'px';
            this.pointer.style.left = x + 'px';

            x = x / 200;
            y = y / 200 - 1;

            x = -x > 0 ? -x : x;
            y = -y > 0 ? -y : y;

            this.s = parseFloat(x.toFixed(2));
            this.l = parseFloat(y.toFixed(2));

            this.change();
        }
    },

    stopDrag: function (e) {
        if (this.pointerMove === true) {
            this.triggerChange();
        }
        this.pointerMove = false;
    },

    hslRgb: function (h, s, v) {
        var r, g, b, i, f, p, q, t;

        i = Math.floor(h * 6);
        f = h * 6 - i;
        p = v * (1 - s);
        q = v * (1 - f * s);
        t = v * (1 - (1 - f) * s);

        switch (i % 6) {
            case 0:
                r = v, g = t, b = p;
                break;
            case 1:
                r = q, g = v, b = p;
                break;
            case 2:
                r = p, g = v, b = t;
                break;
            case 3:
                r = p, g = q, b = v;
                break;
            case 4:
                r = t, g = p, b = v;
                break;
            case 5:
                r = v, g = p, b = q;
                break;
        }

        return [
            Math.round(r * 255),
            Math.round(g * 255),
            Math.round(b * 255)
        ];
    },

    rgbHsl: function (red, green, blue) {
        var rr, gg, bb,
            r = arguments[0] / 255,
            g = arguments[1] / 255,
            b = arguments[2] / 255,
            h, s,
            v = Math.max(r, g, b),
            diff = v - Math.min(r, g, b),
            diffc = function (c) {
                return (v - c) / 6 / diff + 1 / 2;
            };

        if (diff === 0) {
            h = s = 0;
        } else {
            s = diff / v;
            rr = diffc(r);
            gg = diffc(g);
            bb = diffc(b);

            if (r === v) {
                h = bb - gg;
            } else if (g === v) {
                h = (1 / 3) + rr - bb;
            } else if (b === v) {
                h = (2 / 3) + gg - rr;
            }
            if (h < 0) {
                h += 1;
            } else if (h > 1) {
                h -= 1;
            }
        }

        return [parseFloat(h.toFixed(3)), parseFloat(s.toFixed(3)), parseFloat(v.toFixed(3))];
    },

    rgbHex: function (rgb) {
        var hex = [rgb[0].toString(16), rgb[1].toString(16), rgb[2].toString(16)];

        hex.forEach(function (v, k) {
            if (v.length === 1) {
                hex[k] = '0' + v;
            }
        });

        return '#' + hex.join('');
    },

    hexRgb: function (hex) {
        let shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
        hex = hex.replace(shorthandRegex, function (m, r, g, b) {
            return r + r + g + g + b + b;
        });

        let result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);

        return result ? [parseInt(result[1], 16), parseInt(result[2], 16), parseInt(result[3], 16)] : null;
    },


    wrap: function (item) {
        let wrapper = document.createElement('div'),
            thumb = document.createElement('div'),
            value = item.value;
        wrapper.className = 'clpi-wrapper';
        thumb.className = 'clpi-thumb';

        if (value.length) {
            thumb.style.background = value;
        }

        item.parentNode.appendChild(wrapper);

        wrapper.appendChild(item);

        return wrapper.appendChild(thumb);
    },

    initAll: function () {
        for (var item of document.querySelectorAll('.clpi:not(.clpi-initialized)')) {
            item.classList.add('clpi-initialized');

            this.wrap(item);
        }
    },

    setup: function () {
        let that = this;

        this.hslInput.addEventListener('input', function (e) {
            that.setHsl(e.target.value);
        });

        this.hslInput.addEventListener('change', function (e) {
            that.triggerChange();
        });

        this.alphaRange.addEventListener('input', function (e) {
            that.setAlpha(e.target.value);
        });

        this.alphaRange.addEventListener('change', function (e) {
            that.triggerChange();
        });

        this.currentColor.addEventListener('input', function (e) {
            that.setCurrent(e.target.value);
        });

        this.gradientAngleInut.addEventListener('input', function (e) {
            that.setGradientAngle(e.target.value);
        });

        this.area.addEventListener('mousedown', function (e) {
            that.startDrag(e);
        });

        this.mask.addEventListener('mousedown', function (e) {
            if (e.target.matches('#clpi-m')) {
                that.close();
            }
        });

        document.addEventListener('mousemove', function (e) {
            that.drag(e);
        });

        document.addEventListener('mouseup', function (e) {
            that.stopDrag(e);
        });

        document.addEventListener('click', function (e) {
            if (e.target.matches('.clpi') || e.target.matches('.clpi-thumb')) {
                colorPicker.init(e.target);
            }

            if (e.target.matches('.clpi-o')) {
                colorPicker.setGradientControl(e.target);
            }

            if (e.target.matches('.clpi-u')) {
                colorPicker.gradientShowToggle();
            }
        });

        this.initAll();
    },

};

colorPicker.setup();

src/helpers/show_selected_code.ts
import * as vscode from 'vscode';
import getFileName from './get_file_name';

export function showSelectedCodeAsReadme(selectedText: string): void {
    const panel = vscode.window.createWebviewPanel(
      'Code Scenery', // Unique identifier for the panel
      'Code Scenery', // Title displayed in the panel's header
      vscode.ViewColumn.One, // The column in which the panel should be shown
      {
        enableScripts: false // Enable JavaScript and other scripts in the webview
      }
    );
  
    // Set the HTML content of the webview panel
    panel.webview.html = getWebViewContent(selectedText);
  }


  function getWebViewContent(selectedText: string): string {
    // 1) Extract Selected Code
    const codeSnippet = escapeHtml(selectedText);
  
    // 2) Extract File name
    const fileName = getFileName();
  
    // 3) Split code snippet into lines
    const lines = codeSnippet.split('\n');
  
    // 4) Add line numbers to each line
    const codeWithLineNumbers = lines
      .map((line, index) => {
        const lineNumber = index + 1;
        return `<div class="line"><span class="line-number">${lineNumber}</span>${line}</div>`;
      })
      .join('');
  
    return `
      <!DOCTYPE html>
      <html>
  
      <head>
          <!-- Bootstrap CSS -->
          <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/css/bootstrap.min.css"
              integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css">
          <link rel="preconnect" href="https://fonts.googleapis.com">
          <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
          <link href="https://fonts.googleapis.com/css2?family=Cairo:wght@700&display=swap" rel="stylesheet">
      </head>
  
      <body>
          <div class="background">
              <!--========== Start of controls options ================= -->
              <div class="controls-options d-flex">
                  <div class="item">
                      <button type="button" class="btn camera-btn"><i class="fas fa-camera camera-icon"></i>
                      </button>
                  </div>
              </div>
              <!--========== End of controls options ================= -->
              <div class="window-frame">
                  <div class="frame-header">
                      <div class="window-button red"></div>
                      <div class="window-button yellow"></div>
                      <div class="window-button grey"></div>
                      <div class="code-language">${fileName}</div>
                  </div>
                  <div class="code-box">
                      ${codeWithLineNumbers}
                  </div>
              </div>
          </div>
          </div>
          <style>
              @import url('https://fonts.googleapis.com/css2?family=Cairo:wght@700&display=swap');
  
              * {
                  margin: 0;
                  padding: 0;
                  box-sizing: border-box;
                  font-family: 'Cairo', sans-serif;
              }
  
              body {
                  display: flex;
                  flex-direction: column;
                  justify-content: center;
                  align-items: center;
                  width: 100%;
                  height: 100vh;
                  background-image: linear-gradient(140deg, #EADEDB 0%, #BC70A4 50%, #BFD641 75%);
              }
  
              .controls-options {
                  display: flex;
                  justify-content: center;
                  align-items: center;
              }
  
              .item {
                  margin: 1rem;
              }
  
              .camera-btn {
                  width: 5rem;
                  height: 5rem;
                  border-radius: 50%;
                  background-color: transparent;
              }
  
              .camera-icon {
                  font-size: 60px;
              }
  
              .window-frame {
                  position: relative;
                  display: inline-block;
                  background-color: #282828;
                  border-radius: 8px;
                  overflow: hidden;
                  box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.2);
              }
  
              .frame-header {
                  padding: 10px;
                  display: flex;
                  flex-direction: row;
                  align-items: center;
              }
  
              /* Styles for the window buttons */
              .window-button {
                  width: 12px;
                  height: 12px;
                  border-radius: 50%;
                  display: inline-block;
                  margin: 0 5px;
              }
  
              .window-button.red {
                  background-color: #ff5f57;
              }
  
              .window-button.yellow {
                  background-color: #ffbd2e;
              }
  
              .window-button.grey {
                  background-color: #c7c7c7;
              }
  
              /* Styles for the code box */
              .code-box {
                  background-color: #1a1a1a;
                  color: #fff;
                  font-size: 14px;
                  padding: 20px;
                  border-radius: 8px;
                  overflow: auto;
                  position: relative;
              }
  
              /* Styles for the code language */
              .code-language {
                  background-color: #000;
                  color: #fff;
                  padding: 5px 10px;
                  border-radius: 4px;
                  font-weight: bold;
                  margin-bottom: 10px;
                  margin-left: auto;
              }
  
              /* Styles for line numbers */
              .line-number {
                  display: inline-block;
                  text-align: right;
                  padding-right: 10px;
                  color: #808080;
                  user-select: none;
                  cursor: default;
              }
  
              .line {
                  display: flex;
                  align-items: baseline;
              }
  
              code {
                  display: block;
                  padding: 4px 8px;
                  white-space: pre-wrap;
              }
          </style>
  
          <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js"
              integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN"
              crossorigin="anonymous"></script>
          <script src="https://cdn.jsdelivr.net/npm/popper.js@1.12.9/dist/umd/popper.min.js"
              integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q"
              crossorigin="anonymous"></script>
          <script src="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/js/bootstrap.min.js"
              integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl"
              crossorigin="anonymous"></script>
      </body>
  
      </html>
    `;
  }
  


  function escapeHtml(text: string): string {
    // Helper function to escape special characters in HTML
    return text.replace(/[<>&]/g, function (tag) {
      return tag === '<' ? '&lt;' : tag === '>' ? '&gt;' : '&amp;';
    });
  }